[Product Collection] Add Price Range Filter

imanish003 opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

Adds a Price Range Filter to the Product Collection Block's Inspector Controls
<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/62e10296-ec53-49ad-a3ed-a47fbd628b5d" width="300">


- **Integration of Price Range Filter**: A `priceRange` attribute has been added to the block's attribute, allowing merchants to define a price range for displaying products. This is particularly useful for creating more targeted and specific product displays based on pricing criteria.
- **Back-end Support and Filtering Logic**: Corresponding back-end adjustments in `ProductCollection.php` ensure that the product query accounts for the new price range filter, effectively filtering products within the specified price boundaries.
- **Use of 'Auto' for Price Values**: The implementation allows for dynamic price range selection with an 'Auto' placeholder. When the price fields are left blank (auto), the filter dynamically adjusts the range to include all products, offering flexibility in how the price range is applied.

Fixes woocommerce/woocommerce#42377 

## Why

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->
This update significantly improves the versatility of the Product Collection block, empowering merchants to curate products more effectively based on price ranges, and thus enhancing the shopping experience for customers.


## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Insert a product collection block into a new post.
2. From the Filters dropdown, select the “Price Range” filter. It will appear as shown:

<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5a9e6a14-4aa1-4c00-abbf-6c3280524f2a" width="300">

3. Adjust the MIN & MAX values. Ensure that the products displayed in the Editor are filtered according to these updated values.
    - **Note:** The changes will only take effect if you *hit ENTER* or *click elsewhere*. This is designed to prevent unnecessary API calls while the merchant is adjusting the value.
4. Save the modifications and confirm that the same updates are visible on the Frontend.
5. Ensure that you can reset the filter's value by clicking in two different places, as indicated in the screenshot below:
<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/d6f2bf7f-0335-4f15-a28b-b618963a5e7a" width="300">

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [x] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Product Collection - Add Price Range Filter

imanish003 committed:
Add Price Range Filter to Product Collection Block

- Introduced 'priceRange' attribute in the DEFAULT_QUERY and DEFAULT_FILTERS within `constants.ts` to initialize the new price range filter.
- Added new SCSS rules in `editor.scss` for styling the price range control.
- Implemented `PriceRangeControl` in `inspector-controls/index.tsx` to enable users to specify a price range for product filtering.
- Created `PriceTextField.tsx` and `index.tsx` under `inspector-controls/price-range-control`, providing the UI for entering minimum and maximum price values.
- Extended the `ProductCollectionQuery` and `PriceRange` interfaces in `types.ts` to type-check the new price range functionality.
- Updated `ProductCollection.php` to handle the price range data when constructing the product query, ensuring products are filtered based on the specified price range.

imanish003 committed:
Merge branch 'trunk' into 9360-product-collection-filters-price-range

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,48 @@
+
+interface PriceTextFieldProps {
+	value: number;
+	onChange: ( min: number ) => void;
```
Component is used for both: min and max so I think it should be agnostic:

```suggestion
	onChange: ( value: number ) => void;
```

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,48 @@
+		<NumberControl
+			value={ value }
+			onChange={ ( min: string ) => {
+				onChange( Number( min ) );
```
Sam as above: Component is used for both: min and max so I think it should be agnostic:

```suggestion
			onChange={ ( value: string ) => {
				onChange( Number( value ) );
```

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/price-range-control/index.tsx:
```diff
@@ -0,0 +1,76 @@
+			} }
+			className="wc-block-product-price-range-control"
+		>
+			<Text>{ __( 'PRICE RANGE', 'woo-gutenberg-products-block' ) }</Text>
```
What's preventing us from relying on `ToolsPanelItem`'s label? Is the problem that `NumberControl` has it's own `label` and this one is not displayed, so we need to rely on `Text` component?
The reason I'm asking is the visual difference, Price Range stands out among other filters. Maybe we should style it so it's similar to other filter labels.
<img width="293" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/b9bc97b3-195a-4e47-81fa-4144228c0e60">

imanish003 replied:
You are right. Label for `ToolsPanelItem` is only shown in dropdown menu, it doesn't show up with the Filter itself. For other filters, we are using Label with input fields itself, For example label prop with `RangeControl`, `ToggleControl` etc.
But in this case, we don't wanna assign the label to input fields itself (NumberControl), but to the section, therefore I am using `Text` component. Let me try to make it better.

imanish003 replied:
After looking at [this code](https://github.com/WordPress/gutenberg/blob/9905301c5f783141177f61e37851eb9e7e96b875/packages/components/src/utils/base-label.ts#L6-L13) in GB repo, I was able to figure out that we can use `BaseControl.VisualLabel` for the label.

kmanijak reviewed (changes requested):
I'm playing with the Price Range and here are some issues:

### 🔴 Filter doesn't work correctly

Please check the video below, where products with price €54 are filter out when setting MIN value at €46 (if I set min in euro... 🤔).
https://github.com/woocommerce/woocommerce-blocks/assets/20098064/c8207f67-db8a-41de-a1b5-b3948e465735

Another video with max, filtering out product with price €108.90 is filtered out with MAX value €89

https://github.com/woocommerce/woocommerce-blocks/assets/20098064/a68b1684-bf6b-4b50-a212-e73d06f6d8d0

I suspected some currency clash, but my main currency is EURO, so all should be fine:
<img width="702" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/037d522b-f89f-4829-ab61-c8df37b559a0">


### 🟠 Lack of currency sign

As a user, I'd definitely expect the currency sign next to the inputs. That's especially important for the users using multi-currency so they know what currency they set up the limits in.

Here are some additional smaller issues/edge cases. Let me know what you think of them and if it's something you think is worth addressing or we accept that for now:

### Min greater than Max

It's possible to insert `min` value greater than `max`. Product Collection behaves fine, showing "No Results" and obviously it's a merchant's responsibility to make sure it's configured properly.

So probably it's not a must-have but wanted to mention anyway:

<img width="298" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/efb764d7-a490-4c29-89ae-65039d4cf9e7">

### "-" sign

It's possible to insert minus "-" sign in the numbers but the number has no effect then. I know we're resuing the input and not controlling its validation so again - mentioning for the record
<img width="228" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/5015052e-64a7-4541-9b7a-f3ad9c858c04">

### Can't apply decimals

You can input decimals but once submitted, the value is rounded to full number. I think we can give merchants a possibility to set decimals, especially that it should operate on any currency and I guess there are currencies that operate on decimals much more.

kmanijak commented on assets/js/blocks/product-collection/editor.scss:
```diff
@@ -24,3 +24,13 @@
+
+// Price Range Filter
+.wc-block-product-price-range-control {
+	max-width: 200px;
```
What's the purpose of this limit? I think we could let the controls occupy the full Inspector Controls width as it stands out now from other filters:
<img width="290" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/618b8ab5-c340-4d6d-bf67-3c74ea7172b2">

imanish003 replied:
It wasn’t looking great when using full width. But I think it makes sense to remove the width for now & later if required, we can add it again. After adding the currency sign, I think it would look better 🤝

imanish003 commented:
> As a user, I'd definitely expect the currency sign next to the inputs. That's especially important for the users using multi-currency so they know what currency they set up the limits in.

Agreed! I will add the currency sign. Although, AFAIK In Editor, it will always use the base currency which is set using Settings → General → Currency options. Right?

imanish003 commented:
@kmanijak 
> **Min greater than Max**

I agree that we can add validation, but I am not 100% sure how this functionality should work. For example, while editing the MAX value:

- Should we show a message whenever MIN > MAX?
- Should we discard the input whenever MIN > MAX?

Also, as you said, it doesn’t have any consequences because, in such cases, no products will be returned.

imanish003 commented:
> **"-" sign**

Good catch! Although it doesn’t have any issues related to functionality, I will check if I can quickly fix it because I am using [NumberControl](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-numbercontrol--docs), so I think this is something which should be handled by NumberControl itself.

imanish003 commented:
> You can input decimals but once submitted, the value is rounded to full number. I think we can give merchants a possibility to set decimals, especially that it should operate on any currency and I guess there are currencies that operate on decimals much more.

Good catch! I will look into it. Thanks 🙏

imanish003 committed:
Enhanced Price Range Control with Currency Symbol and Flexible Width

- Removed the `max-width` restriction from `.wc-block-product-price-range-control` in `editor.scss`.

- Introduced currency symbol support in `PriceTextField.tsx` by importing `getCurrency` from `@woocommerce/price-format`. The currency symbol is now displayed as a suffix in the price input fields, providing a clearer context for the prices being entered.

- Implemented `step="any"` attribute in `NumberControl` within `PriceTextField.tsx`. This modification allows for more precise price inputs, accommodating prices with decimal values.

imanish003 committed:
Merge branch 'trunk' into 9360-product-collection-filters-price-range

imanish003 commented:
> Filter doesn't work correctly

I haven't been able to replicate the issue yet, but I'll give it another try while working on the other feedback you provided. In the meantime, it would be really helpful if you could attempt to reproduce the issue on a new site.

kmanijak commented:
> I haven't been able to replicate the issue yet, but I'll give it another try while working on the other feedback you provided. In the meantime, it would be really helpful if you could attempt to reproduce the issue on a new site.

I figured out what's the issue.

In WooCommerce -> Settings -> Tax this was my (unrecommended) setup:

<img width="710" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/47484d5c-7e03-49cc-b9b0-0d3d9501a340">

I input prices exclusive of tax and display them including tax which means filtering was happening on the base price (exclusive of tax). I _think_ that maybe the filtering should always happen based on the _displayed_ prices (no matter if inclusive or exclusive of tax) so there are no misunderstandings by merchants similar to me who thought it was a bug 😅. What do you think?

For clarity - after unifying the prices filter works great. I also confirmed on another page with unified setup.

imanish003 committed:
Add support for Decimal values

imanish003 committed:
Make label same as other filter labels

imanish003 committed:
Refactor PriceRangeControl for Improved Layout Consistency

imanish003 commented:
Hi @kmanijak 👋 

Thank you so much for your valuable feedback on this PR! I've taken action on most of the points you raised:

- Added currency sign
- Removed the width from CSS
- Decimal value support is now included.
- The "Price Range" label now matches the style of other labels.

However, I haven't yet addressed the following feedback:

> I input prices exclusive of tax and display them including tax which means filtering was happening on the base price (exclusive of tax). I *think* that maybe the filtering should always happen based on the *displayed* prices (no matter if inclusive or exclusive of tax) so there are no misunderstandings by merchants similar to me who thought it was a bug 😅. What do you think?

This issue is a bit complex, so I've sought some additional input (p1701251026307649-slack-C02FL3X7KR6).

Regarding the "Min greater than Max" and the "-" sign concerns, could you please review the comments I left on these links?

- https://github.com/woocommerce/woocommerce-blocks/pull/11931#issuecomment-1829228028
- https://github.com/woocommerce/woocommerce-blocks/pull/11931#issuecomment-1829231997

Once again, thank you so much for your detailed feedback. Your keen eye for detail during the PR review is greatly appreciated! 🙌🏻

kmanijak commented:
### Currency sign
> Agreed! I will add the currency sign. Although, AFAIK In Editor, it will always use the base currency which is set using Settings → General → Currency options. Right?

I think so, that's my understanding 👍 

### Min greater than Max
> I agree that we can add validation, but I am not 100% sure how this functionality should work. For example, while editing the MAX value:
> - Should we show a message whenever MIN > MAX?
> - Should we discard the input whenever MIN > MAX?
> Also, as you said, it doesn’t have any consequences because, in such cases, no products will be returned.

I'm against using messages. I was thinking that another value should be adjusted to the one you input:
- min: 10, max: 20. You change min to 30, so both are updated to 30
- min: 10, max: 20. You change max to 5, so both are updated to 5
But TBH I don't think that's necessary until we introduce a Range Control. Then we'd need to apply that validation, for now I think it's alright as-is.

### Minus sign
> Although it doesn’t have any issues related to functionality, I will check if I can quickly fix it because I am using [NumberControl](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-numbercontrol--docs), so I think this is something which should be handled by NumberControl itself.

Yes, let's ignore it. It's not important problem to solve at all and actually why would someone input minus in the middle of the number? 😆

imanish003 committed:
Update ProductCollection to handle tax-adjusted price filters

The changes primarily focus on the following areas:
1. **Tax Adjustment Incorporation**: The logic now includes an adjustment for tax considerations. This is facilitated by the newly added `should_adjust_price_range_for_taxes()` method, which determines if the price filters need to be adjusted based on the discrepancy between price storage and display methods concerning tax inclusion.

2. **Tax Class Handling**: A new method `adjust_price_range_for_tax_classes()` has been implemented. It adjusts the price range for each tax class, ensuring accurate and relevant calculations. This method is essential for handling various tax scenarios and classes, improving the accuracy of price range filtering.

3. **Price Adjustment Logic**: The `adjust_price_value_for_tax_class()` method has been added for detailed tax adjustments. This method adjusts price values based on tax classes and the tax display settings, using logic akin to `wc_get_price_excluding_tax` and `wc_get_price_including_tax` from WooCommerce core.

4. **Meta Query Construction**: A new method `build_meta_query_for_price_range()` constructs the meta query for filtering products within the specified price range. This method simplifies and consolidates the meta query construction process.

Overall, these changes aim to make the price filtering feature more robust, especially in scenarios involving different tax setups. The update enhances the module's flexibility and accuracy in handling diverse pricing scenarios.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9360-product-collection-filters-price-range

imanish003 committed:
Fix build error

imanish003 commented:
I am closing this pull request because I have created the same pull request on the WooCommerce repository. Here is the link: https://github.com/woocommerce/woocommerce-blocks/pull/11931.
